### PR TITLE
AEJ-1584 - Fix JNI Exception handling mechanism in GKL

### PR DESCRIPTION
**Pull request by swepnadoss**
AEJ-1584 - Fix JNI Exception handling mechanism in GKL for Deflater, Inflater, PairHMM and SmithWaterman native classes and handle new exceptions propagated from corresponding native code in C classes.

### CODE REVIEW

File: src/main/native/compression/IntelDeflater.cc [hunk 1]
```diff
@@ -90,8 +90,11 @@ JNIEXPORT void JNICALL Java_com_intel_gkl_compression_IntelDeflater_resetNative
     if (lz_stream == 0) {
       lz_stream = (isal_zstream*)malloc(sizeof(isal_zstream));
       if ( lz_stream == NULL ) {
-        jclass Exception = env->FindClass("java/lang/Exception");
-        env->ThrowNew(Exception,"Memory allocation error");
+
+        if(env->ExceptionCheck())
+            env->ExceptionClear();
+
```
Comment (SnehalA) [inline src/main/native/compression/IntelDeflater.cc:96]: Do we expect any exception here?

File: src/main/native/compression/IntelDeflater.cc [hunk 2]
```diff
@@ -121,8 +124,11 @@ JNIEXPORT void JNICALL Java_com_intel_gkl_compression_IntelDeflater_resetNative
     if (lz_stream == 0) {
       lz_stream = (z_stream*)calloc(1, sizeof(z_stream));
       if ( lz_stream == NULL ) {
-        jclass Exception = env->FindClass("java/lang/Exception");
-        env->ThrowNew(Exception,"Memory allocation error");
+
+        if(env->ExceptionCheck())
+            env->ExceptionClear();
+
```
Comment (SnehalA) [inline src/main/native/compression/IntelDeflater.cc:130]: same as above.

File: src/main/native/compression/IntelDeflater.cc [hunk 3]
```diff
@@ -133,8 +139,11 @@ JNIEXPORT void JNICALL Java_com_intel_gkl_compression_IntelDeflater_resetNative
                              nowrap ? -MAX_WBITS : MAX_WBITS,
                              DEF_MEM_LEVEL, Z_DEFAULT_STRATEGY);
       if (ret != Z_OK) {
-        jclass Exception = env->FindClass("java/lang/Exception");
-        env->ThrowNew(Exception,"IntelDeflater init error");
+
+        if(env->ExceptionCheck())
+            env->ExceptionClear();
+
```
Comment (SnehalA) [inline src/main/native/compression/IntelDeflater.cc:145]: can remove extra lines

File: src/main/native/compression/IntelInflater.cc [hunk 4]
```diff
@@ -78,8 +78,11 @@ JNIEXPORT void JNICALL Java_com_intel_gkl_compression_IntelInflater_resetNative
 
      lz_stream = (inflate_state*)calloc(1,sizeof(inflate_state));
       if ( lz_stream == NULL ) {
-        jclass Exception = env->FindClass("java/lang/Exception");
-        env->ThrowNew(Exception,"Memory allocation error");
+
+        if(env->ExceptionCheck())
+            env->ExceptionClear();
+
```
Comment (SnehalA) [inline src/main/native/compression/IntelInflater.cc:84]: same as above

File: src/main/native/compression/IntelDeflater.cc (hunk 1 above)
Comment (swepnadoss) [reply src/main/native/compression/IntelDeflater.cc:96]: As discussed during design/code reviews, we might expect exceptions from malloc and calloc operations like L#91 so this exception check is to handle that.

File: src/main/native/compression/IntelDeflater.cc (hunk 2 above)
Comment (swepnadoss) [reply src/main/native/compression/IntelDeflater.cc:130]: Same as above, check L#125 for calloc

File: src/main/native/compression/IntelInflater.cc (hunk 4 above)
Comment (swepnadoss) [reply src/main/native/compression/IntelInflater.cc:84]: Same as above, we might expect exceptions from malloc and calloc operations like L#79 so this exception check is to handle that.

File: src/main/native/compression/IntelDeflater.cc (hunk 3 above)
Comment (swepnadoss) [reply src/main/native/compression/IntelDeflater.cc]: Done!

**Review by Kmannth — APPROVED** [review]
These changes look and are what we discussed.

**Pull request merged**